File: console/cpp/binary-ante/src/binary.cpp
#include "binary.h"

using namespace std;


/*binary::binary(){
    type = bin8;
    setValue(0);
}*/
binary::binary(BinaryType type, int dec){
    this->type = type;
    setValue(dec);
}
binary::binary(int dec){
    type = bin16;
    setValue(dec);
}
binary::binary(BinaryType type){
    this->type = type;
    setValue(randRange(0, pow(2, (int)type)));
}


int binary::randRange(int min, int max){
    return min + (rand() % (max+1));
}


void binary::setType(BinaryType type){
    this->type = type;
}
void binary::setValue(int dec){
    int max = pow(2, (int)type);
    if(dec < max){
        value = dec;
    } else value = dec % (max-1);
}


int binary::toDec(){
    return value;
}

string binary::toString(){
    string sBin = "";

    int number = value;
    while(number > 0){
        sBin = to_string(number%2) + sBin;
        number /= 2;
    }
    while(sBin.length() < (int)type) sBin = "0" + sBin;

    return sBin;
}

File: console/CSharp/binary-ante/Binary.cs
using System;

namespace binary_ante
{
	public enum BinaryType { bin4=4, bin8=8, bin12=12, bin16=16, bin20=20, bin24=24, bin28=28, bin32=32, bin36=36, bin40=40, bin44=44, bin48=48 };


	public class Binary
	{
		private BinaryType Type;
		private int Value;


		public Binary(){
			Type = BinaryType.bin8;
			SetValue(RandRange(0, (int)Math.Pow(2, (int)Type)));
		}
		public Binary(BinaryType type){
			this.Type = type;
			SetValue(RandRange(0, (int)Math.Pow(2, (int)type)));
		}
		public Binary(int value){
			int b = 4;
			while (b <= 48) {
				if (value < (int) Math.Pow(2, b))
					break;
				
				b += 4;
			}
			Type = (BinaryType) b;
			Value = value;
		}
		public Binary(BinaryType type, int decValue){
			this.Type = type;
			SetValue(decValue);
		}

		public static Binary operator+(Binary lhs, Binary rhs){
			Binary result = new Binary(lhs.Value + rhs.Value);

			return result;
		}


		public void SetValue(int decValue){
			int max = (int)Math.Pow(2, (int)Type);
			if(decValue < max){
				Value = decValue;
			} else Value = decValue % (max-1);
		}


		int RandRange(int min, int max){
			return new Random().Next(min, max);
		}

		public int ToDec(){
			return Value;
		}
		override public string ToString(){
			string s = "";

			int num = Value;
			while (num >= 1) {
				s = (num % 2).ToString() + s;
				num /= 2;
			}
			while (s.Length < (int) Type)
				s = "0" + s;

			return s;
		}
	}
}

